### PR TITLE
WebService improvements: add treatments endpoint, caching for SGVs

### PR DESCRIPTION
**Pull request by jwoglom**
This PR improves the functionality of the xDrip web service by adding a new treatments endpoint, which outputs identically to the Nightscout treatments route. It also adds aggressive caching for treatments and SGVs queried over the web service.

I have been using these changes on my forked version of xDrip for many months, and the caching results in substantial speed improvements. With these changes, the [Urchin Pebble watchface](https://github.com/jwoglom/urchin-cgm) can interface directly with xDrip locally, instead of needing to query Nightscout via the public internet. The most common requests made to the web service (for loading the SGV graph and obtaining treatments) load in sub-100 milliseconds, while before this change they could take longer if there were significant amounts of data.

Querying the new treatments endpoint:
```
$ adb forward tcp:17580 tcp:17580
$ time curl localhost:17580/treatments.json?count=100|jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 17728  100 17728    0     0   141k      0 --:--:-- --:--:-- --:--:--  141k
[
  {
    "_id": "3dd5fdd7-aee1-4ee4-8f0b-d77cbbb1dfb7",
    "created_at": 1619206024253,
    "eventType": "<none>",
    "enteredBy": "xdrip",
    "notes": "example note",
    "carbs": 0,
    "insulin": 0
  },
  ...
]

real	0m0.157s
user	0m0.057s
sys	0m0.015s
```

Querying for 300 sensor glucose values at once:

```
$ time curl localhost:17580/sgv.json?count=300|jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 84143  100 84143    0     0   439k      0 --:--:-- --:--:-- --:--:--  439k
[
  {
    "_id": "aa0af10e-2282-4f00-b64c-9c88d27dbec5",
    "device": "Follower",
    "dateString": "2021-05-01T16:58:06.076-0400",
    "sysTime": "2021-05-01T16:58:06.076-0400",
    "date": 1619902686076,
    "sgv": 148,
    "delta": -7,
    "direction": "FortyFiveDown",
    "noise": 1,
    "filtered": 0,
    "unfiltered": -127,
    "rssi": 100,
    "type": "sgv",
    "units_hint": "mgdl"
  },
  ...
]

real    0m0.225s
user    0m0.078s
sys     0m0.034s
```

### CODE REVIEW

**Comment by jamorham** [comment]
Assuming this works and doesn't introduce any bugs or incompatibilities with existing behavior then I think this is a good addition.

File: app/src/main/java/com/eveningoutpost/dexdrip/localeTasker/receiver/FireReceiver.java [hunk 1]
```diff
@@ -162,8 +162,14 @@ public void onReceive(final Context context, final Intent intent) {
                             break;
 
                         case "SNOOZE":
-                            AlertPlayer.defaultSnooze();
-                            JoH.static_toast_long("SNOOZE from Tasker");
+                            int minutes = -1; // default
+                            if (message_array.length > 1) {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/localeTasker/receiver/FireReceiver.java:166]: If this change is not related to the intend of the PR, please remove it.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceProfile.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * @author James Woglom (j@wogloms.net)
+ */
+
+public class WebServiceProfile extends BaseWebService {
+
+    private static String TAG = "WebServiceProfile";
+
+    public WebResponse request(String query) {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceProfile.java:13]: What does this class do? The `query` parameter is not used inside the `request` method.

**Review by tolot27 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/eveningoutpost/dexdrip/localeTasker/receiver/FireReceiver.java (hunk 1 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/localeTasker/receiver/FireReceiver.java]: Done. Removed the tasker related changes -- rebased and removed the commit. I'll publish a separate CR with this change and some other tasker additions.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceProfile.java (hunk 2 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceProfile.java]: Removed this class and related profile route code.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.eveningoutpost.dexdrip.wearintegration.ExternalStatusService.getLastStatusLine;
+import static com.eveningoutpost.dexdrip.wearintegration.ExternalStatusService.getLastStatusLineTime;
+
+
+/**
+ * Created by jwoglom
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java:31]: Please either use valid javadoc tags (`@author`) or leave the author out. Your name is logged with the commit. See [Stop using Javadoc @author tag](https://www.vojtechruzicka.com/stop-using-javadoc-author-tag) for the background.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java [hunk 4]
```diff
@@ -106,8 +112,48 @@ public WebResponse request(String query) {
...
+        // readings is significantly slower, this optimizes the most often use case.
+        List<BgReading> bgr = BgReading.latest(1, ignore_sensor);
+        BgReading latestReading = null;
+        if (bgr != null) {
+            Iterator<BgReading> it = bgr.iterator();
+            latestReading = it.hasNext() ? it.next() : null;
+        }
+
+        List<BgReading> readings;
+        if (this.cachedReadings != null && latestReading != null &&
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:127]: Same logic as in `WebServiceTreatments.java`.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java [hunk 5]
```diff
@@ -106,8 +112,48 @@ public WebResponse request(String query) {
...
         final boolean ignore_sensor = Home.get_follower() || cgi.containsKey("all_data");
 
-        final List<BgReading> readings = BgReading.latest(count, ignore_sensor);
+
+        // Store a cache of the last BgReading.latest() query for the duration in which there is no
+        // new latest reading. Since obtaining the latest reading is fast, but a larger number of
+        // readings is significantly slower, this optimizes the most often use case.
+        List<BgReading> bgr = BgReading.latest(1, ignore_sensor);
+        BgReading latestReading = null;
+        if (bgr != null) {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:121]: Could be

```java
        if (bgr != null && bgr.size() > 0) {
            latestReading = bgr.iterator().next();
        }
```

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java [hunk 6]
```diff
@@ -106,8 +112,48 @@ public WebResponse request(String query) {
...
+                }
+            }
+        } else {
+            UserError.Log.d(TAG, "Fetching latest "+count+" readings from BgReading");
+            readings = BgReading.latest(count, ignore_sensor);
+            this.cachedReadings = readings;
+        }
+
+        // Prevent the cache from growing too large. 1000 entries is about 3.5 days
+        if (cachedJson.size() > 1000) {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:149]: Isn't it more efficient to use a `LinkedHashMap` and implement `removeEldestEntry` in such a way that the [oldest entries are removed automatically](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html#removeEldestEntry-java.util.Map.Entry-), if the map contains 1.000 entries already and a new element gets added, instead of removing all entries?

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java [hunk 7]
```diff
@@ -184,9 +235,12 @@ public WebResponse request(String query) {
                         sensor_status_string = null;
                     }
 
+                    cachedJson.put(reading.uuid, item);
+                    withManual++;
                     reply.put(item);
                 }
 
+                Log.d(TAG, "Processed "+withCache+" cached entries and "+withManual+" manual entries");
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:243]: Why not use `UserError.Log.d` here as well?

**Review by tolot27 — CHANGES_REQUESTED** [review]
Thanks again for your effort and cleanup. I've revised the two bigger change sets.

In general: apply code formatting and avoid double-empty lines.

And please resolve the conversation if you are ready.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java (hunk 6 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java]: Clever solution -- done.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java (hunk 7 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java]: Done

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java (hunk 3 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Done. This was done automatically by a really old version of intellij

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+        int count = 24;
+
+        if (cgi.containsKey("count")) {
+            try {
+                count = Integer.valueOf(cgi.get("count"));
+                count = Math.min(count, 100);
+                count = Math.max(count, 1);
+                UserError.Log.d(TAG, "Treatment count request for: " + count + " entries");
+            } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Done

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                // meh
+            }
+        }
+
+        final JSONArray reply = new JSONArray();
```
[the remark this answers is not in the document]
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Done

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+
+        // Store a cache of the last Treatments.latest() query for the duration in which there is no
+        // new latest treatment. Since obtaining the latest treatment is fast, but a larger number of
+        // treatments is significantly slower, this optimizes the most often use case.
+        Iterator<Treatments> it = Treatments.latest(1).iterator();
```
[the remark this answers is not in the document]
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Done

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        Treatments latestTreatment = it.hasNext() ? it.next() : null;
+
+        List<Treatments> treatments;
+        if (this.cachedTreatments != null && latestTreatment != null &&
+            this.cachedTreatments.size() > 0 && latestTreatment.uuid.equals(this.cachedTreatments.iterator().next().uuid) &&
+            count <= this.cachedTreatments.size())
```
[the remark this answers is not in the document]
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Done

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        {
+            if (count == this.cachedTreatments.size()) {
+                UserError.Log.d(TAG, "Using cached treatments");
+                treatments = this.cachedTreatments;
+            } else {
+                UserError.Log.d(TAG, "Copying "+count+" of "+this.cachedTreatments.size()+" cached treatments");
```
[the remark this answers is not in the document]
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Done. In my day job I'm very used to having an autoformatter, but I've fixed all of the inconsistent code styles I could find

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+                    treatments.add(it.next());
+                }
+            }
+        } else {
+            UserError.Log.d(TAG, "Fetching latest "+count+" entries from Treatments");
+            treatments = Treatments.latest(count);
+            this.cachedTreatments = treatments;
+        }
+
+        if (treatments != null) {
```
[the remark this answers is not in the document]
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Done

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 14]
```diff
@@ -53,63 +48,65 @@ public WebResponse request(String query) {
                 count = Math.min(count, 100);
                 count = Math.max(count, 1);
                 UserError.Log.d(TAG, "Treatment count request for: " + count + " entries");
-            } catch (Exception e) {
-                // meh
+            } catch (NumberFormatException e) {
+                UserError.Log.wtf(TAG, "Invalid treatment count request for: " + count + " entries");
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java:52]: Hmm, wtf is not appropriate here, IMHO.

**Comment by tolot27** [comment]
Many thanks for the changes. Looks like you have some dev experience. May I ask you to create a test case for your implementation?

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (tolot27) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: I've enabled code auto formatting, too. We should add this to the dev/contributor docs.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceModule.java [hunk 15]
```diff
@@ -19,69 +21,77 @@
 @Module
 public class WebServiceModule {
 
-    private static final boolean d = true;
+    private static final String TAG = "WebServiceModule-INJECT";
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceModule.java:24]: Hopefully, this changed log entry name does not affect any downstream log analyses. Including the class name is indeed usefull.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java [hunk 16]
```diff
@@ -100,23 +110,58 @@ public WebResponse request(String query) {
...
+        List<BgReading> bgr = BgReading.latest(1, ignore_sensor);
+        BgReading latestReading = null;
+        if (bgr != null && bgr.size() > 0) {
+            latestReading = bgr.iterator().next();
+        }
+
+        List<BgReading> readings;
+        if (cachedReadings != null && cachedReadings.size() > 0 && count <= cachedReadings.size() &&
+            latestReading != null && latestReading.uuid.equals(cachedReadings.iterator().next().uuid))
+        {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:130]: The `{` could be at the end of the previous line. :-o

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java [hunk 17]
```diff
@@ -140,56 +187,69 @@ public WebResponse request(String query) {
...
-                    }
+        // Add special fields for first SGV entry
+        if (reply.length() > 0) {
+            try {
+                // Copy the first item only to a new JSONObject as to not
+                // add these items to the cached json object
+                JSONObject origItem = reply.getJSONObject(0);
+                JSONObject item = new JSONObject(origItem.toString());
+                if (units_indicator > 0) {
+                    String unitsHint = Pref.getString("units", "mgdl").equals("mgdl") ? "mgdl" : "mmol";
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:213]: What should this check do? Should it ensure that either "mgdl" or "mmol" is returned? I thought only either of them is allowed in the preferences.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+            latestTreatment = latestList.iterator().next();
+        }
+
+        // Check whether there are any cached treatments, and if so whether the latest cached
+        // treatment matches the UUID of the most recent treatment. Use the cache if we can,
+        // otherwise fill cachedTreatments with new treatments from the database.
+        List<Treatments> treatments;
+        if (cachedTreatments != null && cachedTreatments.size() > 0 && count <= cachedTreatments.size() &&
+            latestTreatment != null && latestTreatment.uuid.equals(cachedTreatments.iterator().next().uuid))
+        {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java:71]: Small formatting change again: move `{` to previous line.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+                item.put("notes", treatment.notes);
+                item.put("carbs", treatment.carbs);
+                item.put("insulin", treatment.insulin);
+
+                reply.put(item);
+            }
+
+            Log.d(TAG, "Treatment output: " + reply.toString());
+        } catch (JSONException e) {
+            UserError.Log.wtf(TAG, "Got json exception: " + e);
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java:109]: Again, not a terrible failure which breaks xDrip, just the webservice response.

**Review by tolot27 — CHANGES_REQUESTED** [review]
Just some minor changes. Looks really good so far. A unit test would be great!

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java (hunk 14 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Changed to Log.w (warning)

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java (hunk 18 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Done

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java (hunk 19 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java]: Changed to w

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java (hunk 16 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java]: Done

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java (hunk 17 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:212]: Honestly looks like some kind of sanity check for if the preference didn't exist. I copied this from earlier in the method, doesn't seem incredibly worth changing since it doesn't hurt anything.

**Comment by jwoglom** [comment]
> Many thanks for the changes. Looks like you have some dev experience. May I ask you to create a test case for your implementation?

I've added some more robust tests in RouteFinderTest.java

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java (hunk 17 above)
Comment (tolot27) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:212]: The declaration is `String getString(final String pref, final String def)`. Only if the preference is not set, it returns "mgdl". If the preferences could not be initialized, an empty string is returned by `Pref.getString`. Hence, always "mmol" is returned with the exception of "mgdl".

**Review by tolot27 — CHANGES_REQUESTED** [review]
Great test coverage so far. Unfortunately, caching is not really covered. Please can you add it, too?

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java [hunk 20]
```diff
@@ -100,23 +110,57 @@ public WebResponse request(String query) {
...
+        // readings is significantly slower, this optimizes the most often use case.
+        List<BgReading> bgr = BgReading.latest(1, ignore_sensor);
+        BgReading latestReading = null;
+        if (bgr != null && bgr.size() > 0) {
+            latestReading = bgr.iterator().next();
+        }
+
+        List<BgReading> readings;
+        if (cachedReadings != null && cachedReadings.size() > 0 && count <= cachedReadings.size() &&
+            latestReading != null && latestReading.uuid.equals(cachedReadings.iterator().next().uuid)) {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:129]: Line 129-140 are not covered.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+        if (latestList != null && !latestList.isEmpty()) {
+            latestTreatment = latestList.iterator().next();
+        }
+
+        // Check whether there are any cached treatments, and if so whether the latest cached
+        // treatment matches the UUID of the most recent treatment. Use the cache if we can,
+        // otherwise fill cachedTreatments with new treatments from the database.
+        List<Treatments> treatments;
+        if (cachedTreatments != null && cachedTreatments.size() > 0 && count <= cachedTreatments.size() &&
+            latestTreatment != null && latestTreatment.uuid.equals(cachedTreatments.iterator().next().uuid)) {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java:70]: Line 70-82 are not covered.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java (hunk 20 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:134]: Added tests which check for the presence of WebServiceSgv.cachedReadings and WebServiceSgv.cachedJson

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java (hunk 21 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceTreatments.java:74]: Added tests which check for the presence of WebServiceTreatments.cachedTreatments

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java (hunk 20 above)
Comment (tolot27) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:134]: Caching itself is still not covered but likely harder to implement a test case for it.

**Review by tolot27 — APPROVED** [review]
A great PR with a very high test coverage which improves the WebService a lot by caching SGVs and adds support for treatments.

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java [hunk 22]
```diff
@@ -140,56 +191,69 @@ public WebResponse request(String query) {
...
+                // TODO uploader battery
+
+                // emit sensor status/age message if present
+                if (sensor_status_string != null) {
+                    item.put("sensor_status", sensor_status_string);
+                }
+                reply.put(0, item);
             } catch (JSONException e) {
-                UserError.Log.wtf(TAG, "Got json exception: " + e);
+                e.printStackTrace();
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:256]: Shouldn't we inform the user if this exception occurs?

**Review by jamorham — APPROVED** [review]

File: app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java (hunk 22 above)
Comment (jwoglom) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:256]: Eh, this would be a programming bug if it ever occurred. Shouldn't actually be possible for this to happen, the JSON put method just explicitly requires that the JSONException be caught.

Comment (tolot27) [reply app/src/main/java/com/eveningoutpost/dexdrip/webservices/WebServiceSgv.java:256]: But `e.printStackTrace()` is logged to adb, AFAIK. Hence, you will only see this exception during debugging. Isn't it better to use `UserError.Log.e()` instead?
